### PR TITLE
Search: Prepare application for Gutenberg customization

### PR DESCRIPTION
**Pull request by jsnmoon**
#### Changes proposed in this Pull Request:

Prepares the Jetpack Search Preact application for the upcoming Gutenberg customization interface. Specifically:

* Adds the ability to disable analytics integration.
* Adds the ability to disable DOM integration (mainly related to reading and manipulating the current window's query string).
* Adds support for handling mutable overlay options while preserving support for the WordPress Customizer.
* Registers each search customization option as a [WordPress site setting](https://developer.wordpress.org/reference/functions/register_setting/).
* Refactors initial state generation in PHP into a helper function (`generate_initial_javascript_state`).
* Modularizes `@automattic/color-studio` literalization in Webpack into a separate file. 

#### Jetpack product discussion
pcNPJE-7J-p2

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
* Apply this change to your Jetpack site with a Search subscription. Alternatively, use Jurassic Ninja to spin up a test site for this branch.
* Perform a search on the site.
  * Ensure that the analytics calls fire as expected; look for `t.gif` requests with a defined `blog_id` query string value.
  * Ensure that the window's query string updates as expected; it should add `?s=` when opening the search modal, for example.
* Customize Jetpack Search options in the Customizer.
  * Ensure that the overlay spawns upon opening the Jetpack Search section of the Customizer.
  * Ensure that each Customization option correctly updates the search overlay. For default sort and excluded post types, the customization iframe should refresh to reflect your changes.

* (bonus) Ensure that `pnpm build` works as expected.

### CODE REVIEW

File: projects/plugins/jetpack/modules/search/instant-search/components/search-app.jsx [hunk 1]
```diff
@@ -45,17 +47,35 @@ import './search-app.scss';
...
+
 		this.state = {
-			overlayOptions: { ...this.props.initialOverlayOptions },
 			isVisible: !! this.props.initialIsVisible, // initialIsVisible can be undefined
+			overlayOptionsCustomizerOverride: {},
 		};
+
 		this.getResults = debounce( this.getResults, 200 );
-		this.props.initializeQueryValues();
+		this.props.enableAnalytics && this.initializeAnalytics();
```
Comment (kangzj) [inline projects/plugins/jetpack/modules/search/instant-search/components/search-app.jsx:63]: `initializeAnalytics` is not called only when `enableAnalytics` is set to false, as expected . But unfortunately, the tracking is still enabled somehow - I can still see tracking data sent, e.g.
`https://pixel.wp.com/t.gif?fetch_algo=jetpack:search/1-score_default&fetch_position=2&fetch_query=hello&railcar=dmHMQ0jwdjjb&rec_blog_id=193141071&rec_post_id=44&session_id=TGN6e2&ui_algo=jetpack-instant-search-ui/v1&ui_position=2&_en=jetpack_instant_search_traintracks_render&_ui=Aywbd8KXVwkYLpcKh7BhG1O9&_ut=anon&_ts=1624492779267&_tz=-12&_lg=en-US&_pf=MacIntel&_ht=1120&_wd=1792&_sx=0&_sy=0&_dl=https://jasper.au.ngrok.io/?s=hello&_dr=&_rt=1624492779267&_=_`

**Review by kangzj — CHANGES_REQUESTED** [review]
Good job Jason 👍 
The analytics data is always sent regardless of the value of `enableAnalytics`. Other parts all work as proposed.

----
EDIT: Seems other modules would initialize and use analytics as well. So probably we'll have just to set a switch on our `recordEvent` function.

File: projects/plugins/jetpack/modules/search/instant-search/components/search-app.jsx (hunk 1 above)
Comment (jsnmoon) [reply projects/plugins/jetpack/modules/search/instant-search/components/search-app.jsx:63]: Drat -- I thought I sorted this out. I'll look into this.

File: projects/plugins/jetpack/modules/search/instant-search/lib/tracks.js [hunk 2]
```diff
@@ -1,26 +1,67 @@
...
 }
 
+/**
+ * Fires a TrainTracks render event to Tracks.
+ *
+ * @param {object} properties - Event properties.
+ */
 export function recordTrainTracksRender( properties ) {
-	recordEvent( 'jetpack_instant_search_traintracks_render', properties );
+	isAnalyticsEnabled && recordEvent( 'jetpack_instant_search_traintracks_render', properties );
```
Comment (kangzj) [inline projects/plugins/jetpack/modules/search/instant-search/lib/tracks.js:55]:
```suggestion
	recordEvent( 'jetpack_instant_search_traintracks_render', properties );
```
`recordEvent` already tested whether analytics is enabled or not.

File: projects/plugins/jetpack/modules/search/instant-search/lib/tracks.js [hunk 3]
```diff
@@ -1,26 +1,67 @@
...
 }
 
+/**
+ * Fires a TrainTracks interaction event to Tracks.
+ *
+ * @param {object} properties - Event properties.
+ */
 export function recordTrainTracksInteract( properties ) {
-	recordEvent( 'jetpack_instant_search_traintracks_interact', properties );
+	isAnalyticsEnabled && recordEvent( 'jetpack_instant_search_traintracks_interact', properties );
```
Comment (kangzj) [inline projects/plugins/jetpack/modules/search/instant-search/lib/tracks.js:64]:
```suggestion
	recordEvent( 'jetpack_instant_search_traintracks_interact', properties );
```

File: projects/plugins/jetpack/modules/search/instant-search/lib/tracks.js [hunk 4]
```diff
@@ -29,5 +70,5 @@ export function recordTrainTracksInteract( properties ) {
  * @param {object} properties - Event properties to send to Tracks.
  */
 export function recordStaticFilterSelect( properties ) {
-	recordEvent( 'jetpack_instant_search_static_filter_select', properties );
+	isAnalyticsEnabled && recordEvent( 'jetpack_instant_search_static_filter_select', properties );
```
Comment (kangzj) [inline projects/plugins/jetpack/modules/search/instant-search/lib/tracks.js:73]:
```suggestion
	recordEvent( 'jetpack_instant_search_static_filter_select', properties );
```

**Review by kangzj — DISMISSED** [review]
Looks good to me. Good job Jason 👍

File: projects/plugins/jetpack/modules/search/instant-search/lib/tracks.js (hunk 2 above)
Comment (jsnmoon) [reply projects/plugins/jetpack/modules/search/instant-search/lib/tracks.js]: 👍

**Review by jeherve — CHANGES_REQUESTED** [review]
This is going to need a rebase I'm afraid.

**Review by jeherve — APPROVED** [review]
This tests well for me. 🚢

**Comment by sdixon194** [comment]
Just a headsup @jsnmoon that the wpcom patch couldn't be applied cleanly. I forced the test to pass so we could merge this PR, but could be worth double checking the wpcom PR!

**Comment by jsnmoon** [comment]
Thanks @sdixon194! I'd already updated D63247-code manually, but will double-check before deploying just in case :)

**Comment by jsnmoon** [comment]
🚢 r227910-wpcom

**Pull request merged**